README.md
Kapost Bridge Logger
=================
[![Build Status](https://travis-ci.org/webbuilders-group/silverstripe-kapost-bridge-logger.png)](https://travis-ci.org/webbuilders-group/silverstripe-kapost-bridge-logger)

A wrapper for our SilverStripe Kapost Bridge module that logs all incoming requests to the kapost service.

## Maintainer Contact
* Ed Chipman ([UndefinedOffset](https://github.com/UndefinedOffset))

## Requirements
* SilverStripe CMS 3.1+
* [Kapost Bridge](https://github.com/webbuilders-group/silverstripe-kapost-bridge)


## Installation
__Composer (recommended):__
```
composer require webbuilders-group/silverstripe-kapost-bridge-logger
```


If you prefer you may also install manually:
* Download the module from here https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/archive/master.zip
* Extract the downloaded archive into your site root so that the destination folder is called kapost-bridge-logger, opening the extracted folder should contain _config.php in the root along with other files/folders
* Run dev/build?flush=all to regenerate the manifest


## Configuration Options
```yml
KapostBridgeLog:
    log_expire_days: 30 #Number of days that logs are kept

KapostBridgeLogViewer:
    log_page_length: 20 #Number of logs to include per page

LoggedKapostService:
    ignore_not_found: true #Whether to log 404 errors for example when the request is not post or user agent checking is enabled and does not match
```

## Extension Points
There is one extension point that allows you to hook into the lookup process for the destination object. This extension point is called ``updateObjectLookup`` and is on the KapostBridgeLog class. The extension point is given one argument which is the Kapost Reference ID. Extensions using this extension point should return an object (or null) with the CMSEditLink method defined if you want a link to appear on the log. The CMSEditLink method must return a url to the edit page for that object. The first extension to return an object is used.


## Documentation
Additional documentation for the module can be found [here](docs/en).


CHANGELOG.md
# Change Log

## [0.3.2](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.3.2) (2017-05-11)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.3.1...0.3.2)

## [0.3.1](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.3.1) (2017-05-03)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.3.0...0.3.1)

## [0.3.0](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.3.0) (2017-01-12)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.2.4...0.3.0)

**Implemented enhancements:**

- Refresh Button for Log list [\#6](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/issues/6)

## [0.2.4](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.2.4) (2016-10-24)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.2.3...0.2.4)

## [0.2.3](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.2.3) (2016-06-20)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.2.2...0.2.3)

## [0.2.2](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.2.2) (2016-06-06)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.2.1...0.2.2)

## [0.2.1](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.2.1) (2016-04-20)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.2.0...0.2.1)

## [0.2.0](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.2.0) (2016-04-13)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.1.5...0.2.0)

**Implemented enhancements:**

- Move logger under the Kapost Content manager [\#4](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/issues/4)

**Merged pull requests:**

- Added standard Scrutinizer config [\#5](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/pull/5) ([helpfulrobot](https://github.com/helpfulrobot))

## [0.1.5](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.1.5) (2016-01-29)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.1.4...0.1.5)

**Merged pull requests:**

- Added standard .editorconfig file [\#3](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/pull/3) ([helpfulrobot](https://github.com/helpfulrobot))

## [0.1.4](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.1.4) (2015-10-13)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.1.3...0.1.4)

## [0.1.3](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.1.3) (2015-10-05)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.1.2...0.1.3)

## [0.1.2](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.1.2) (2015-09-30)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.1.1...0.1.2)

**Implemented enhancements:**

- Improve copying of the log content [\#1](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/issues/1)

## [0.1.1](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.1.1) (2015-09-23)
[Full Changelog](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/compare/0.1.0...0.1.1)

## [0.1.0](https://github.com/webbuilders-group/silverstripe-kapost-bridge-logger/tree/0.1.0) (2015-09-09)


tests/KapostBridgeLoggerTest.php
<?php
class KapostBridgeLoggerTest extends FunctionalTest {
    const USER_AGENT='Kapost XMLRPC::Client';
    
    protected static $fixture_file='KapostBridgeLoggerTest.yml';
    
    
    /**
     * Test that 404's are ignored
     */
    public function testIgnorePageNotFound() {
        //Try just a get
        $response=$this->get('/kapost-service');
        $this->assertEquals(404, $response->getStatusCode());
        $this->assertEquals(0, KapostBridgeLog::get()->count());
        
        
        //Try just a post
        $response=$this->post('/kapost-bridge', array());
        $this->assertEquals(404, $response->getStatusCode());
        $this->assertEquals(0, KapostBridgeLog::get()->count());
    }
    
    /**
     * Test to see that the request/response was actaully logged
     */
    public function testBridgeLogging() {
        $response=$this->call_service('get-post');
        
        //Ensure we had a 200 response
        $this->assertEquals(200, $response->getStatusCode());
        
        
        //Find the log
        $log=KapostBridgeLog::get()->sort('ID DESC')->first();
        
        
        //Ensure it exists
        $this->assertNotEmpty($log);
        $this->assertNotEquals(false, $log);
        $this->assertTrue($log->exists());
        
        
        //Verify the log has the request details
        $this->assertNotEmpty($log->Request);
        
        
        //Verify the log has the response details
        $this->assertNotEmpty($log->Response);
        
        
        //Verify the called method is correct
        $this->assertEquals('metaWeblog.getPost', $log->Method);
    }
    
    
    /**
     * Tests to verify the password was stripped from the request
     */
    public function testPasswordStripping() {
        $response=$this->call_service('get-post');
        
        //Ensure we had a 200 response
        $this->assertEquals(200, $response->getStatusCode());
        
        
        //Find the log
        $log=KapostBridgeLog::get()->sort('ID DESC')->first();
        
        
        //Ensure it exists
        $this->assertNotEmpty($log);
        $this->assertNotEquals(false, $log);
        $this->assertTrue($log->exists());
        
        
        //Parse the XML
        $xml=simplexml_load_string($log->Request);
        
        $this->assertEquals('['._t('LoggedKapostService.PASSWORD_FILTERED', '_PASSWORD FILTERED').']', $xml->params->param[2]->value->string);
    }
    
    /**
     * Tests to verify that the bits of the asset were stripped from the request
     */
    public function testBitsStripping() {
        $response=$this->call_service('new-media-asset');
        
        //Ensure we had a 200 response
        $this->assertEquals(200, $response->getStatusCode());
        
        
        //Find the log
        $log=KapostBridgeLog::get()->sort('ID DESC')->first();
        
        
        //Ensure it exists
        $this->assertNotEmpty($log);
        $this->assertNotEquals(false, $log);
        $this->assertTrue($log->exists());
        
        
        //Parse the XML
        $xml=simplexml_load_string($log->Request);
        
        $this->assertEquals('['._t('LoggedKapostService.BITS_FILTERED', '_BASE64 BITS FILTERED').']', $xml->params->param[3]->value->struct->member[2]->value->base64);
    }
    
    /**
     * Calls the api and returns the response
     * @param string $mockRequest Mock Request to load
     * @return SS_HTTPResponse Response Object
     */
    protected function call_service($mockRequest) {
        return $this->post('kapost-service', array(), array('User-Agent'=>self::USER_AGENT), null, file_get_contents(dirname(__FILE__).'/mock_requests/'.$mockRequest.'.xml'));
    }
    
    /**
     * Parses the response from the api
     * @param string $body XML Response
     * @return PhpXmlRpc\Request XML RPC Response Object
     */
    final protected function parseRPCResponse($body) {
        $xmlmsg=new PhpXmlRpc\Request('');
        
        return $xmlmsg->parseResponse($body, true, 'phpvals');
    }
}
?>

code/model/KapostBridgeLog.php
<?php
/**
 * Class KapostBridgeLog
 *
 * @property string $Method
 * @property string $Request
 * @property string $Response
 * @property string $UserAgent
 * @mixin LenovoKapostBridgeLog
 */
class KapostBridgeLog extends DataObject {
    private static $db=array(
                            'Method'=>'Varchar(100)',
                            'Request'=>'Text',
                            'Response'=>'Text',
                            'UserAgent'=>'Text'
                         );
    
    private static $default_sort='Created DESC, ID DESC';
    
    /**
     * Time in days to expire the log entries
     * @config KapostBridgeLog.log_expire_days
     * @var int
     * @default 30
     */
    private static $log_expire_days=30;
    
    private static $casting=array(
                                'RequestFormatted'=>'Text',
                                'ResponseFormatted'=>'Text'
                            );
    
    private $reference_object=false;
    
    
    /**
     * Gets the title of this log which is a combination of the method called and the date/time
     * @return string Title of the log entry
     */
    public function getTitle() {
        return $this->Method.': '.$this->dbObject('Created')->FormatFromSettings();
    }
    
    /**
     * Cleans up logs older than X days after writing
     */
    protected function onAfterWrite() {
        parent::onAfterWrite();
        
        //Clean up old logs
        $oldLogs=KapostBridgeLog::get()->filter('Created:LessThan', date('Y-m-d H:i:s', strtotime('-'.self::config()->log_expire_days.' days')));
        if($oldLogs->count()>0) {
            foreach($oldLogs as $log) {
                $log->delete();
            }
        }
    }
    
    /**
     * Gets the request formatted with line breaks
     * @return string Raw XML Request from Kapost
     */
    public function getRequestFormatted() {
        $request=$this->Request;
        if(!empty($request)) {
            $doc=new DomDocument('1.0');
            $doc->preserveWhiteSpace=false;
            $doc->formatOutput=true;
            if(!@$doc->loadXML($request)) {
                return _t('KapostBridgeLogViewer.REQUEST_PARSE_ERROR', '_Could not Parse Request');
            }
            
            return $doc->saveXML();
        }
    }
    
    /**
     * Gets the response formatted with line breaks
     * @return string Raw XML Response from SilverStripe
     */
    public function getResponseFormatted() {
        $response=$this->Response;
        if(!empty($response)) {
            $doc=new DomDocument('1.0');
            $doc->preserveWhiteSpace=false;
            $doc->formatOutput=true;
            if(!@$doc->loadXML($response)) {
                return _t('KapostBridgeLogViewer.RESPONSE_PARSE_ERROR', '_Could not Parse Response');
            }
            
            return $doc->saveXML();
        }
    }
    
    /**
     * Gets the referenced object of this kapost bridge log entry
     */ 
    public function getReferenceObject() {
        if($this->reference_object===false) {
            $kapostRefID=false;
            
            if($this->Method=='metaWeblog.getPost' || $this->Method=='metaWeblog.editPost') {
                $xml=simplexml_load_string($this->Request);
                if($xml) {
                    $kapostRefID=$xml->params->param[0]->value->string->__toString();
                }
            }else if($this->Method=='metaWeblog.newPost' || $this->Method=='kapost.getPreview' || $this->Method=='metaWeblog.newMediaObject') {
                $xml=simplexml_load_string($this->Response);
                if($xml && !isset($xml->fault)) {
                    if($this->Method=='metaWeblog.newPost') {
                        $kapostRefID=$xml->params->param[0]->value->string->__toString();
                    }else if($this->Method=='kapost.getPreview') {
                        $kapostRefID=$xml->params->param[0]->value->struct->member[1]->value->string->__toString();
                    }else if($this->Method=='metaWeblog.newMediaObject') {
                        //Attempt to lookup the referenced file
                        $fileURL=$xml->params->param[0]->value->struct->member[1]->value->string->__toString();
                        if(!empty($fileURL)) {
                            $file=KapostService::find_file_by_url($fileURL);
                            if(!empty($file) && $file!==false && $file->ID>0) {
                                $this->reference_object=$file;
                                return $this->reference_object;
                            }else {
                                $this->reference_object=null;
                                return;
                            }
                        }else {
                            $this->reference_object=null;
                            return;
                        }
                    }
                }
            }else {
                $this->reference_object=null;
                return;
            }
            
            
            //If we have a 
            if(!empty($kapostRefID)) {
                Versioned::reset(); //Reset versioned to remove filters
                
                //Try looking for a page
                $page=SiteTree::get()->filter('KapostRefID', Convert::raw2sql($kapostRefID))->first();
                if(!empty($page) && $page!==false && $page->exists()) {
                    $this->reference_object=$page;
                }else {
                    //Allow extensions to add their own logic
                    $extensionResult=array_filter($this->extend('updateObjectLookup', $kapostRefID), function($item) {return is_object($item);});
                    $extensionResult=array_shift($extensionResult);
                    if(!empty($extensionResult) && $extensionResult!==false && $extensionResult->exists()) {
                        $this->reference_object=$extensionResult;
                    }else {
                        //Look for a kapost object that is not a preview
                        $kapostObj=KapostObject::get()->filter('KapostRefID', Convert::raw2sql($kapostRefID))->first();
                        if(!empty($kapostObj) && $kapostObj!==false && $kapostObj->exists() && $kapostObj->IsKapostPreview==false) {
                            $this->reference_object=$kapostObj;
                        }else {
                            //No result so set to null
                            $this->reference_object=null;
                        }
                    }
                }
            }else {
                //No Kapost Reference ID so set to null
                $this->reference_object=null;
            }
        }
        
        return $this->reference_object;
    }
    
    /**
     * Gets the link to view this log entry
     * @return string
     */
    public function CMSEditLink() {
        return Controller::join_links(LeftAndMain::config()->url_base, KapostBridgeLogViewer::config()->url_segment, 'view', $this->ID);
    }
}
?>

code/forms/KapostLogLinkField.php
<?php
class KapostLogLinkField extends FormField {
    protected $allowHTML=true;
    
    protected $_link;
    protected $_label;
    
    /**
     * Constructor
     * @param string $name Name of the field
     * @param string $title Title for use on the field
     * @param string $link Link target to use
     * @param string $label Link label to use
     */
    public function __construct($name, $title=null, $link=null, $label=null) {
        parent::__construct($name, $title);
        
        $this->_link=$link;
        $this->_label=$label;
    }
    
    /**
     * Sets the label to use in the link
     * @param string $value
     */
    public function setLinkLabel($value) {
        $this->_label=$value;
        return $this;
    }
    
    /**
     * Gets the label to used in the link
     * @return string
     */
    public function getLinkLabel() {
        return $this->_label;
    }
    
    /**
     * Sets the target of the link
     * @param string $value
     */
    public function setLinkTarget($value) {
        $this->_link=$value;
        return $this;
    }
    
    /**
     * Gets the target of the link
     * @return string
     */
    public function getLinkTarget() {
        return $this->_link;
    }
    
    /**
     * Generates the contents of the field
     * @param array $properties Properties to add to the field (not used)
     * @return string
     */
    public function Field($properties=null) {
        if(!empty($this->_link) && !empty($this->_label)) {
            return '<span id="'.$this->ID().'" class="readonly"><a href="'.Convert::raw2att($this->_link).'" class="cms-panel-link">'.Convert::raw2xml($this->_label).'</a></span>';
        }
    }
    
    /**
     * Generates the normal field holder for the field
     * @param array $properties Properties to add to the field holder
     * @return string
     */
    public function FieldHolder($properties=null) {
        if(!empty($this->_link) && !empty($this->_label)) {
            return parent::FieldHolder($properties);
        }
    }
    
    /**
     * Generates the small field holder for the field
     * @param array $properties Properties to add to the field holder
     * @return string
     */
    public function SmallFieldHolder($properties=null) {
        if(!empty($this->_link) && !empty($this->_label)) {
            return parent::SmallFieldHolder($properties);
        }
    }
    
    /**
     * This field has no data so return false
     * @return bool
     */
    public function hasData() {
        return false;
    }
    
    /**
     * Returns a readonly version of this field
     */
    public function performReadonlyTransformation() {
        $clone=clone $this;
        $clone->setReadonly(true);
        
        return $clone;
    }
    
    /**
     * Type of the field is readonly
     * @return string
     */
    public function Type() {
        return 'readonly';
    }
}
?>

code/control/KapostBridgeLogViewer.php
<?php
/**
 * Class KapostBridgeLogViewer
 *
 */
class KapostBridgeLogViewer extends LeftAndMain implements PermissionProvider {
    private static $url_segment='kapost-bridge-logs';
    private static $tree_class='KapostBridgeLog';
    private static $log_page_length=20;
    
    private static $allowed_actions=array(
                                        'view'
                                    );
    
    private $_menu_updated=false;
    
    
    /**
     * Adds requirements needed for the log viewer
     */
    public function init() {
        parent::init();
        
        Requirements::css('kapost-bridge-logger/css/KapostBridgeLogViewer.css');
        Requirements::javascript('kapost-bridge-logger/javascript/KapostBridgeLogViewer.js');
    }
    
    /**
     * Gets the form used for viewing a time log
     */
    public function getEditForm($id=null, $fields=null) {
        $record=$this->currentPage();
        if($this->action=='view' && $record) {
            $fields=new FieldList(
                                new HeaderField('LogHeader', _t('KapostBridgeLogViewer.VIEWING_ENTRY', '_Viewing Log Entry: {datetime}', array('datetime'=>$record->dbObject('Created')->FormatFromSettings())), 3),
                                new ReadonlyField('UserAgent', _t('KapostBridgeLogViewer.USER_AGENT', '_Requestor User Agent')),
                                new ReadonlyField('Method', _t('KapostBridgeLogViewer.METHOD', '_Method')),
                                ToggleCompositeField::create('RequestData', _t('KapostBridgeLogViewer.KAPOST_REQUEST', '_Kapost Request'), new FieldList(
                                                                                                ReadonlyField::create('RequestFormatted', '')
                                                                                                    ->setTemplate('KapostBridgeLogField')
                                                                                                    ->addExtraClass('log-contents cms-panel-layout')
                                                                                            ))->setHeadingLevel(3),
                                ToggleCompositeField::create('ResponseData', _t('KapostBridgeLogViewer.SILVERSTRIPE_RESPONSE', '_SilverStripe Response'), new FieldList(
                                                                                                ReadonlyField::create('ResponseFormatted', '')
                                                                                                    ->setTemplate('KapostBridgeLogField')
                                                                                                    ->addExtraClass('log-contents cms-panel-layout')
                                                                                            ))->setHeadingLevel(3)
                            );
            
            
            $refObj=$record->ReferenceObject;
            if(!empty($refObj) && $refObj!==false && $refObj->exists()) {
                if(method_exists($refObj, 'CMSEditLink')) {
                    $fields->insertBefore(new KapostLogLinkField('CMSEditLink', _t('KapostBridgeLogViewer.REFERENCED_OBJECT', '_Referenced Object'), $refObj->CMSEditLink(), _t('KapostBridgeLogViewer.VIEW_REFERENCED_OBJECT', '_View Referenced Object')), 'RequestData');
                }else if($refObj instanceof File) {
                    $refObjLink=Controller::join_links(LeftAndMain::config()->url_base, AssetAdmin::config()->url_segment, 'EditForm/field/File/item', $refObj->ID, 'edit');
                    $fields->insertBefore(new KapostLogLinkField('CMSEditLink', _t('KapostBridgeLogViewer.REFERENCED_OBJECT', '_Referenced Object'), $refObjLink, _t('KapostBridgeLogViewer.VIEW_REFERENCED_OBJECT', '_View Referenced Object')), 'RequestData');
                }
            }
        }else {
            $fields=new FieldList();
        }
        
        
        $form=new CMSForm($this, 'EditForm', $fields, new FieldList());
        $form->setResponseNegotiator($this->getResponseNegotiator());
        $form->addExtraClass('cms-edit-form center');
        $form->setAttribute('data-layout-type', 'border');
        $form->setTemplate($this->getTemplatesWithSuffix('_EditForm'));
        $form->setAttribute('data-pjax-fragment', 'CurrentForm');
        $form->setHTMLID('Form_EditForm');
        
        if($record) {
            $form->loadDataFrom($record);
        }
        
        
        return $form;
    }
    
    /**
     * Handles requests to view logs
     * @return mixed Returns PjaxResponseNegotiator if we're using ajax, 404 if we're using ajax and the response cannot be found, redirect if not found in a non-ajax request, and an array if found in an ajax request.
     */
    public function view() {
        //If we're dealing with an ajax request return the form's html
        if(Director::is_ajax()) {
            //If the log cannot be found 404
            if(!$this->currentPage()) {
                return $this->httpError(404);
            }
            
            return $this->getResponseNegotiator()->respond($this->request);
        }
        
        
        //If the log cannot be found redirect to the main screen
        if(!$this->currentPage()) {
            return $this->redirect($this->Link());
        }
        
        
        //Other wise render normally
        return array();
    }
    
    public function MainMenu($cached=true) {
        $menu=parent::MainMenu($cached);
        
        if(!$cached || !$this->_menu_updated) {
            $item=$this->_cache_MainMenu->find('Code', 'KapostAdmin');
            if($item) {
                $item->LinkingMode='current';
            }
        }
        
        return $menu;
    }
    
    /**
     * Current menu item is the KapostAdmin
     * @return ArrayData
     */
    public function MenuCurrentItem() {
        return $this->MainMenu()->find('Code', 'KapostAdmin');
    }
    
    /**
     * Form used for displaying the currently logged items
     */
    public function LogsForm() {
        $fields=new FieldList(
                            DropdownField::create('CalledMethod', _t('KapostBridgeLogViewer.CALLED_METHOD', '_Called Method'), array(
                                                                                    'blogger.getUsersBlogs'=>'blogger.getUsersBlogs',
                                                                                    'kapost.getPreview'=>'kapost.getPreview',
                                                                                    'metaWeblog.editPost'=>'metaWeblog.editPost',
                                                                                    'metaWeblog.getCategories'=>'metaWeblog.getCategories',
                                                                                    'metaWeblog.getPost'=>'metaWeblog.getPost',
                                                                                    'metaWeblog.newMediaObject'=>'metaWeblog.newMediaObject',
                                                                                    'metaWeblog.newPost'=>'metaWeblog.newPost',
                                                                                    'system.listMethods'=>'system.listMethods'
                                                                                ))->setEmptyString('--- '._t('KapostBridgeLogViewer.FILTER_BY_METHOD', '_Filter by Method').' ---'),
                            $startDate=new DatetimeField('LogStartDate', _t('KapostBridgeLogViewer.START_DATE_TIME', '_Start Date/Time')),
                            $endDate=new DatetimeField('LogEndDate', _t('KapostBridgeLogViewer.END_DATE_TIME', '_End Date/Time'))
                        );
        
        
        $startDate->getDateField()->setConfig('showcalendar', true)
                                    ->setAttribute('placeholder', _t('KapostBridgeLogViewer.DATE_FORMAT', '_MMM d, y'))
                                    ->setDescription(_t('KapostBridgeLogViewer.DATE_FORMAT_DESC', '_e.g. Jan 27, 2016'));
        
        $startDate->getTimeField()
                                ->setAttribute('placeholder', _t('KapostBridgeLogViewer.TIME_FORMAT', '_h:mm:ss a'))
                                ->setDescription(_t('KapostBridgeLogViewer.TIME_FORMAT_DESC', '_e.g. 12:15:13 PM'));
        
        
        $endDate->getDateField()->setConfig('showcalendar', true)
                                ->setAttribute('placeholder', _t('KapostBridgeLogViewer.DATE_FORMAT', '_MMM d, y'))
                                ->setDescription(_t('KapostBridgeLogViewer.DATE_FORMAT_DESC', '_e.g. Jan 27, 2016'));
        
        $endDate->getTimeField()
                                ->setAttribute('placeholder', _t('KapostBridgeLogViewer.TIME_FORMAT', '_h:mm:ss a'))
                                ->setDescription(_t('KapostBridgeLogViewer.TIME_FORMAT_DESC', '_e.g. 12:15:13 PM'));
        
        
        $actions=new FieldList(
                                FormAction::create('doApplyFilters', _t('KapostBridgeLogViewer.APPLY_FILTER', '_Apply Filter'))->addExtraClass('ss-ui-action-constructive')->setUseButtonTag(true),
                                Object::create('ResetFormAction', 'clear', _t('KapostBridgeLogViewer.RESET', '_Reset'))->setUseButtonTag(true)
                            );
        
        
        $form=new Form($this, 'LogsForm', $fields, $actions);
        $form->addExtraClass('log-search-form')
            ->setFormMethod('GET')
            ->setFormAction($this->Link())
            ->disableSecurityToken()
            ->unsetValidator();
        
        
        // Load the form with previously sent search data
        $getVars=$this->request->getVars();
        
        //Workaround for start date field with no date or time
        if(array_key_exists('LogStartDate', $getVars)) {
            if(array_key_exists('date', $getVars['LogStartDate']) && !array_key_exists('time', $getVars['LogStartDate'])) {
                $getVars['LogStartDate']['time']='00:00:00';
            }else if(!array_key_exists('date', $getVars['LogStartDate']) && array_key_exists('time', $getVars['LogStartDate'])) {
                unset($getVars['LogStartDate']); //Remove if there is no date present
            }
        }
        
        //Workaround for end date field with no date or time
        if(array_key_exists('LogEndDate', $getVars)) {
            if(array_key_exists('date', $getVars['LogEndDate']) && !array_key_exists('time', $getVars['LogEndDate'])) {
                $getVars['LogEndDate']['time']='23:59:59';
            }else if(!array_key_exists('date', $getVars['LogEndDate']) && array_key_exists('time', $getVars['LogEndDate'])) {
                unset($getVars['LogEndDate']); //Remove if there is no date present
            }
        }
        
        $form->loadDataFrom($getVars);
        
        
        return $form;
    }

	/**
	 * Renders a panel containing the logs available
	 * @return string HTML to be used in the template
	 */
	public function LogsPanel() {
		return $this->renderWith('KapostBridgeLogViewer_Logs');
	}
    
	/**
	 * Gets the logs currently in the database
	 * @return DataList Data List pointing to the logs in the database
	 */
    public function getLogs() {
        $logs=KapostBridgeLog::get();
        
        $filterFields=$this->LogsForm()->Fields();
        
        //Apply Called Method filter
        $var=$filterFields->dataFieldByName('CalledMethod')->Value();
        if(!empty($var)) {
            $logs=$logs->filter('Method', Convert::raw2sql($var));
        }
        
        //Apply Start Date Filter
        $dateTimeField=$filterFields->dataFieldByName('LogStartDate');
        $var=trim($dateTimeField->getDateField()->dataValue().' '.$dateTimeField->getTimeField()->dataValue());
        if(!empty($var)) {
            $logs=$logs->filter('Created:GreaterThan', Convert::raw2sql($var));
        }
        
        //Apply End Date Filter
        $var=$filterFields->dataFieldByName('LogEndDate')->Value();
        if(!empty($var) && $var!=' 00:00:00') {
            $logs=$logs->filter('Created:LessThan', Convert::raw2sql($var));
        }
        
        return PaginatedList::create($logs, $this->request)->setPageLength(self::config()->log_page_length);
    }
    
    /**
     * Gets the pjax response negotiator for this controller
     * @return PjaxResponseNegotiator
     */
    public function getResponseNegotiator() {
        if(!$this->responseNegotiator) {
            parent::getResponseNegotiator();
            
            $controller=$this;
            $this->responseNegotiator->setCallback('LogEntries', function() use(&$controller) {
                                            						return $controller->renderWith('KapostBridgeLogViewer_LogsList');
                                            					});
        }
        
        return $this->responseNegotiator;
    }
    
    public function Breadcrumbs($unlinked=false) {
        $sng=singleton('KapostAdmin');
        $crumbs=new ArrayList(array(
                                    new ArrayData(array(
                                        				'Title'=>$sng->SectionTitle(),
                                        				'Link'=>($unlinked ? false:$sng->Link())
                                        			))
                                ));
        
        $crumbs->merge(parent::Breadcrumbs($unlinked));
        
        return $crumbs;
    }
    
    /**
     * Provides the CMS_ACCESS_KapostBridgeLogViewer permission
     * @return array Map describing the permission for this cms panel
     */
    public function providePermissions() {
        return array(
                    'CMS_ACCESS_KapostBridgeLogViewer'=>array(
                                                            'name'=>_t(
                                                                    'CMSMain.ACCESS',
                                                                    "Access to '{title}' section",
                                                                    "Item in permission selection identifying the admin section. Example: Access to 'Files & Images'",
                                                                    array('title'=>_t('KapostBridgeLogViewer.MENUTITLE', 'Kapost Bridge Logs'))
                                                                ),
                                                            'category'=>_t('Permission.CMS_ACCESS_CATEGORY', 'CMS Access')
                                                        )
                );
    }
}
?>